walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,146 @@
+    // TODO: check for events that all events not crossing EOD.
+
+    if (duration > TimeRange.WHOLE_DAY.duration() || duration <= 0) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private void getOccupiedTimeRange(Set<String> attendees, List<TimeRange> occupiedTimeRange, Event event) {
+    if (Utils.hasIntersection(event.getAttendees(), attendees)) {
+      occupiedTimeRange.add(event.getWhen());
+    }
+  }
+
+  private List<TimeRange> getAvailableTimeRange(List<TimeRange> occupiedTimeRange, long duration) {
+    // Assumes 0 < duration <= WHOLE_DAY
+
+    List<TimeRange> availableMeetings = new ArrayList<>();
+
+    if (occupiedTimeRange.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<TimeRange> mergedTimeRanges = mergeTimeRange(occupiedTimeRange);
+
+    int start = TimeRange.START_OF_DAY;
+
+    for (TimeRange timeRange: mergedTimeRanges) {
+      // Interval duration is not enough.
+      if ((timeRange.start() - start) < duration) {
+        int startForInclusive = Math.min(timeRange.end() + 1, TimeRange.END_OF_DAY);
+        int startForNonInclusive = Math.min(timeRange.end(), TimeRange.END_OF_DAY);
+        start = timeRange.contains(timeRange.end()) ? startForInclusive : startForNonInclusive;
+        continue;
+      }
+
+      TimeRange availableMeeting = TimeRange.fromStartEnd(start, timeRange.start(), /* inclusive= */false);
+      availableMeetings.add(availableMeeting);
+
+      int startForInclusive = Math.min(timeRange.end() + 1, TimeRange.END_OF_DAY);
+      int startForNonInclusive = Math.min(timeRange.end(), TimeRange.END_OF_DAY);
+      start = timeRange.contains(timeRange.end()) ? startForInclusive : startForNonInclusive;
+    }
+
+    // Deals with the last remaining time slot of a day.
+    if (start < TimeRange.END_OF_DAY && (TimeRange.END_OF_DAY - start) >= duration) {
+      availableMeetings.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, /* inclusive= */true));
+    }
+
+    return availableMeetings;
+  }
+
+  private List<TimeRange> mergeTimeRange(List<TimeRange> timeRanges) {
+    List<TimeRange> mergedTimeRanges = new ArrayList<>();
+
+    if (timeRanges.isEmpty()) {
+      return mergedTimeRanges;
+    }
+
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+
+    TimeRange last = null;
+    for (TimeRange timeRange: timeRanges) {
+      // The first TimeRange or no overlapping.
+      if (last == null || !timeRange.overlaps(last)) {
+        mergedTimeRanges.add(timeRange);
+        last = timeRange;
+      } else {
+        TimeRange mergedTimeRange;
+        if (last.contains(timeRange)) {
+          continue;
+        }
+
+        mergedTimeRange = TimeRange.fromStartEnd(last.start(), timeRange.end(), /* inclusive= */false);
+
+        mergedTimeRanges.remove(mergedTimeRanges.size() - 1);
+        mergedTimeRanges.add(mergedTimeRange);
+        last = mergedTimeRange;
+      }
+    }
+
+    return mergedTimeRanges;
```

Comment:
don't return mutable collections. `ImmutableList.copyOf(mergedTimeRanges)`

Reply:
still also for private. keep your return values immutable as much as possible.

Reply:
OK get it:)